pilot/pkg/serviceregistry/kube/controller/endpointslice.go

```diff
@@ -178,6 +178,22 @@ func (esc *endpointSliceController) buildIstioEndpoints(es any, hostName host.Na
 	return esc.endpointCache.Get(hostName)
 }
 
+func endpointHealthStatus(svc *model.Service, e v1.Endpoint) model.HealthStatus {
+	if e.Conditions.Ready == nil || *e.Conditions.Ready {
+		return model.Healthy
+	}
+
+	if features.PersistentSessionLabel != "" &&
+		svc != nil &&
+		svc.Attributes.Labels[features.PersistentSessionLabel] != "" &&
+		(e.Conditions.Serving == nil || *e.Conditions.Serving) &&
+		(e.Conditions.Terminating == nil || *e.Conditions.Terminating) {
+		return model.Draining
```

Comment:
```
	// Ready indicates whether this endpoint is ready and NOT terminating.
	// For pods, this is true if a pod has a ready status and a nil deletion timestamp.
	// This is only set when watching EndpointSlices. If using Endpoints, this is always
	// true since only ready endpoints are read from Endpoints.
	// TODO: Ready can be inferred from Serving and Terminating below when enabled by default.
	Ready bool
	// Serving indiciates whether this endpoint is ready regardless of its terminating state.
	// For pods this is true if it has a ready status regardless of its deletion timestamp.
	// This is only set when watching EndpointSlices. If using Endpoints, this is always
	// true since only ready endpoints are read from Endpoints.
	Serving bool
	// Terminating indicates whether this endpoint is terminating.
	// For pods this is true if it has a non-nil deletion timestamp.
	// This is only set when watching EndpointSlices. If using Endpoints, this is always
	// false since terminating endpoints are always excluded from Endpoints.
	Terminating bool
```